libbpf-tools/biotop.c

```diff
@@ -0,0 +1,392 @@
+	{ "rows", 'r', "ROWS", 0, "Maximum rows to print, default 20" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{ NULL, 'h', NULL, OPTION_HIDDEN, "Show the full help" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	long rows;
+	static int pos_args;
+
+	switch (key) {
+	case 'C':
+		clear_screen = false;
+		break;
+	case 's':
+		if (!strcmp(arg, "all")) {
+			sort_by = ALL;
+		} else if (!strcmp(arg, "io")) {
+			sort_by = IO;
+		} else if (!strcmp(arg, "bytes")) {
+			sort_by = BYTES;
+		} else if (!strcmp(arg, "time")) {
+			sort_by = TIME;
+		} else {
+			warn("invalid sort method: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case 'r':
+		errno = 0;
+		rows = strtol(arg, NULL, 10);
+		if (errno || rows <= 0) {
+			warn("invalid rows: %s\n", arg);
+			argp_usage(state);
+		}
+		output_rows = rows;
+		if (output_rows > OUTPUT_ROWS_LIMIT)
+			output_rows = OUTPUT_ROWS_LIMIT;
+		break;
+	case 'v':
+		verbose = true;
+		break;
+	case 'h':
+		argp_state_help(state, stderr, ARGP_HELP_STD_HELP);
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			interval = strtol(arg, NULL, 10);
+			if (errno || interval <= 0) {
+				warn("invalid interval\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			count = strtol(arg, NULL, 10);
+			if (errno || count <= 0) {
+				warn("invalid count\n");
+				argp_usage(state);
+			}
+		} else {
+			warn("unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_int(int signo)
+{
+	exiting = 1;
+}
+
+static int sort_column(const void *obj1, const void *obj2)
+{
+	struct val_t *s1 = (struct val_t *) obj1;
+	struct val_t *s2 = (struct val_t *) obj2;
+
+	if (sort_by == IO)
+		return s2->io - s1->io;
+	else if (sort_by == BYTES)
+		return s2->bytes - s1->bytes;
+	else if (sort_by == TIME)
+		return s2->us - s1->us;
+	else
+		return (s2->io + s2->bytes + s2->us)
+			- (s1->io + s1->bytes + s1->us);
+}
+
+static void parse_disk_stat(void)
+{
+	FILE *fp;
+	char *line;
+	size_t zero;
+	size_t disk_number;
+
+	fp = fopen("/proc/diskstats", "r");
+	if (fp == NULL)
+		return;
+
+	disk_number = 0;
+	zero = 0;
+	while (getline(&line, &zero, fp) != -1) {
+		struct disk disk;
+
+		if (sscanf(line, "%d %d %s", &disk.major, &disk.minor,
+			disk.name) != 3)
+			continue;
+
+		disks = realloc(disks, (disk_number + 1) * sizeof *disks);
+
+		disks[disk_number] = malloc(sizeof disk);
+		memcpy(disks[disk_number], &disk, sizeof disk);
+
+		disk_number++;
+	}
+
+	free(line);
+	fclose(fp);
+
+	disk_number++;
+	disks = realloc(disks, disk_number * sizeof *disks);
+	disks[disk_number] = NULL;
+}
```

Comment:
I think it'll be simpler if you use a linked list for `disks`, then you won't have to realloc() the array containing pointers and so on.

Reply:
Using a linked list would increase the time of looping through because I will not get access from the cache.
Anyway, I stated [I am not satisfied with this solution](https://github.com/iovisor/bcc/pull/3893#issuecomment-1059718669), so I would like to wait if we decide to switch from C to Rust for this control plane.